### DISCUSSION TITLE
Problem installing 2decomp

### DISCUSSION THREAD

**Discussion by LionelGelebart** [xcompact3d/2decomp-fft#296] [General]
I am trying to install this new version of 2decomp (I am using the old version with a Makefile).
I have followed the instructions to install 2decomp and get the following error with cmake.
I am not familiar with cmake.
Could you help?
-------------------------------------------------------------------------------------------------
[lgelebart@marquises0 src (dev)]$ echo $FC
mpif90
[lgelebart@marquises0 src (dev)]$ cmake --version
cmake version 3.24.0-rc4

CMake suite maintained and supported by Kitware (kitware.com/cmake).
[lgelebart@marquises0 src (dev)]$
[lgelebart@marquises0 src (dev)]$ cmake -S . -B build
CMake Warning (dev) in CMakeLists.txt:
  No project() command is present.  The top-level CMakeLists.txt file must
  contain a literal, direct call to the project() command.  Add a line of
  code such as

    project(ProjectName)

  near the top of the file, but after cmake_minimum_required().

  CMake is pretending there is a "project(Project)" command on the first
  line.
This warning is for project developers.  Use -Wno-dev to suppress it.

-- Compiling using Generic FFT
CMake Error at CMakeLists.txt:93 (install):
  install DIRECTORY given no DESTINATION!

CMake Error at CMakeLists.txt:95 (install):
  install DIRECTORY given no DESTINATION!

CMake Error at CMakeLists.txt:99 (install):
  install EXPORT given no DESTINATION!

CMake Error: File /home/lgelebart/LIBRAIRIES/2decomp-fft/src/cmake/decomp2d-config.cmake.in does not exist.
CMake Error at CMakeLists.txt:102 (configure_file):
  configure_file Problem configuring file

CMake Error: File /home/lgelebart/LIBRAIRIES/2decomp-fft/src/cmake/decomp2d-config-version.cmake.in does not exist.
CMake Error at CMakeLists.txt:106 (configure_file):
  configure_file Problem configuring file

CMake Error: File /home/lgelebart/LIBRAIRIES/2decomp-fft/src/cmake/decomp2d-build_config.cmake.in does not exist.
CMake Error at CMakeLists.txt:119 (configure_file):
  configure_file Problem configuring file

CMake Warning (dev) in CMakeLists.txt:
  No cmake_minimum_required command is present.  A line of code such as

    cmake_minimum_required(VERSION 3.24)

  should be added at the top of the file.  The version specified may be lower
  if you wish to support older CMake versions for this project.  For more
  information run "cmake --help-policy CMP0000".
This warning is for project developers.  Use -Wno-dev to suppress it.

-- Configuring incomplete, errors occurred!
See also "/home/lgelebart/LIBRAIRIES/2decomp-fft/src/build/CMakeFiles/CMakeOutput.log".
[lgelebart@marquises0 src (dev)]$

**Comment by mathrack** [comment]
Hello,

You need to run the cmake command from the main folder, not the SRC one.

Regards,
Mathrack

Nov 15, 2023 12:55:02 LionelGelebart ***@***.***>:

**Reply by LionelGelebart** [reply]
Hello, 
I tried it before with 2 variants: 
1 : cmake -S src -B build
I get the same error message as obtained when running cmake in src
2 : cmake -S . -B build
-> CMake Error: The source "/home/lgelebart/LIBRAIRIES/2decomp-fft/CMakeLists.txt" does not match the source "/home/lgelebart/LIBRAIRIES/2decomp-fft/src/CMakeLists.txt" used to generate cache.  Re-run cmake with a different source directory.

**Reply by LionelGelebart** [reply]
cmake -S . -B build
works after deleting the build.
Sorry (I go on)

**Reply by mathrack** [reply]
You can look at the CI for some examples of the options one can provide to the cmake command
https://github.com/2decomp-fft/2decomp-fft/blob/dev/.github/workflows/Build_all.yml

**Reply by LionelGelebart** [reply]
OK, thank you!

**Reply by LionelGelebart** [reply]
Now it builds against fftw3_f03!
How can I run the differents tests ?

**Comment by mathrack** [comment]
See https://github.com/2decomp-fft/2decomp-fft/blob/dev/.github/workflows/RunnerFullCI.yml#L35

Nov 15, 2023 13:35:20 LionelGelebart ***@***.***>:

**Comment by LionelGelebart** [comment]
OK it works, but only with a single MPI proc. How can I modify the number of MPI process?

**Reply by rfj82982** [reply]
What do you mean a single mpi procs? if this is for running the tests you can simply 
`[bash] ccmake path_to_build`
press t and look for `MPIEXEC_MAX_NUMPROCS` and change to the number of MPI ranks that you have available. 
Indeed if you build the tests you should have a folder `RunTests` and you can simply use 
`mpirun -n XX path_to_build/bin/the_test_i want_to_run `

**Reply by rfj82982** [reply]
Every test takes also a number of inputs if you want to change the resolution, the decomposition and in some the number of iterations

**Reply by rfj82982** [reply]
Test needs to be build and you can do it by 
`[bash] ccmake path_to_build`
and make sure that the option 'BUILD_TESTING' is 'ON'

**Comment by mathrack** [comment]
If you look carefully at the CI examples, you will probably find some stuff for changing the number of grid points and the number of MPI ranks

Nov 15, 2023 14:02:14 LionelGelebart ***@***.***>:

**Comment by pbartholomew08** [comment]
Alternatively you can use `ccmake build/` to configure the tests using a text interface (will require rebuilding). However this only affects the running of the tests - users of the library are free to use however many ranks they choose when they initialise `2decomp&fft`

**Reply by pbartholomew08** [reply]
^ N.B. "affects the running of the tests" using the `ctest` interface, see running generally https://github.com/2decomp-fft/2decomp-fft/discussions/296#discussioncomment-7577029

**Comment by rfj82982** [comment]
Just a clarification when you add the `-S` you need to point at the root of the repository. In theory you could have also created the build under `src` but you should have had 
`cmake -S ../ -b build`
However I'll strongly suggest you not to do this and start to mix the build in the directory where you have also the sources.

**Comment by LionelGelebart** [comment]
Thank you all for your different answers.
1/ When a test individually with mpirun -n XX path_to_build/bin/the_test_i want_to_run, as proposed by rfj82982 : it works correctly and uses the correct number of proc (as observed in the log file)
-> OK!
2/ When I run ctests --test-dir build : tests are running correctly, 0 failed 
BUT, when I look at the log files, only 1 proc is used, even if MAX_NUMPROCS=14 when looking at ccmake build.
I probably do something wrong... any idea?

**Reply by pbartholomew08** [reply]
Hi, `MAX_NUMPROCS` is the most ranks it could run on (it should match the number of cores on your machine). Using `ccmake` or otherwise you need to set `MPIEXEC_MAX_NUMPROCS` which is the number of ranks `ctest` will use.

**Reply by LionelGelebart** [reply]
Yes but MPIEXEC_MAX_NUMPROCS is not present in the 'ccmake build' page (see a copy beow).

                                                     Page 1 of 1
 BUILD_SHARED_LIBS                OFF
 BUILD_TARGET                     mpi
 BUILD_TESTING                    ON
 CMAKE_BUILD_TYPE                 RELEASE
 CMAKE_Fortran_FLAGS_DEV          -DDEBUG -g3 -Og -ffpe-trap=invalid,zero -fcheck=all -fimplicit-none -Wall -Wpedantic -Wno-unused-function -Werror
 CMAKE_INSTALL_PREFIX             /home/lgelebart/LIBRAIRIES/2decomp-fft/build/opt
 DOUBLE_PRECISION                 ON
 ENABLE_INPLACE                   OFF
 ENABLE_PROFILER                  OFF
 EVEN                             OFF
 FFT_Choice                       fftw_f03
 HALO_DEBUG                       OFF
 IO_BACKEND                       mpi
 MAX_NUMPROCS                     14
 NX                               0
 NY                               0
 NZ                               0
 PCOL                             0
 PROW                             0
 SINGLE_PRECISION_OUTPUT          OFF

**Reply by LionelGelebart** [reply]
Haa... it appears when clicking on the [toggle advanced mode].
I try...

**Reply by pbartholomew08** [reply]
I think you can also set it directly during configure (`cmake -DMPIEXEC_MAX_NUM_PROCS=<NP> ...`) just a matter of personal preference

**Reply by LionelGelebart** [reply]
I have just tested this but when using cmake -DMPIEXEC_MAX_NUMPROCS=8, it is MAX_NUMPROCS that is modified in the ccmake build page.
I did this:
rm -rf build/
cmake -S . -B build -DBUILD_TESTING=ON -DMPIEXEC_MAX_NUMPROCS=8
ccmake build

**Reply by LionelGelebart** [reply]
Modifying MPIEXEC_MAX_NUMPROCS  directly through ccmake build is OK (log files show the correct nimber of proc)

**Reply by rfj82982** [reply]
The option is there also in `ccmake` you need just to toggle the advance option and move through the pages (see attached). However, if you do it at configure level it is even better
<img width="764" alt="Screenshot 2023-11-15 at 10 02 26" src="https://github.com/2decomp-fft/2decomp-fft/assets/35581400/613cedd3-fe52-4111-ae95-c24b1efa02dd">

**Reply by LionelGelebart** [reply]
Yes I found it (see 4 messages above). But there is still a remaining problem : see 2 messages above.
Thanks for all

**Reply by mathrack** [reply]
This looks a bit inconsistent. If you think there is a bug in our cmake / ccmake build scripts, feel free to open a issue.
